anonymizer/OpenDiffix.Service/Led.fs

```diff
@@ -43,244 +28,174 @@ let private mergeGivenAggregatorsInto (targetBucket: Bucket) aggIndexes (sourceB
-
   member this.IncrementBucketsIsolatingColumn() =
-    bucketsIsolatingColumn <- bucketsIsolatingColumn + 1
+    Interlocked.Increment(&bucketsIsolatingColumn) |> ignore
 
-  member this.IncrementBucketsMerged() = bucketsMerged <- bucketsMerged + 1
+  member this.IncrementBucketsUnknownColumn() =
+    Interlocked.Increment(&bucketsUnknownColumn) |> ignore
 
-  member this.IncrementComparisons() = comparisons <- comparisons + 1UL
+  member this.IncrementBucketsMerged() =
+    Interlocked.Increment(&bucketsMerged) |> ignore
 
-  member this.Print(bucketsLength: int, victimBucketsLength: int) =
+  member this.IncrementMerge(bucketCount: int) =
+    // Not thread safe!
+    mergeHistogram |> Dictionary.increment bucketCount
+
+  member this.Print(totalBuckets: int, suppressedBuckets: int) =
     let mergeHistogramStr =
       mergeHistogram
       |> Seq.sortBy (fun pair -> pair.Key)
       |> Seq.map (fun pair -> $"({pair.Key}, {pair.Value})")
       |> Seq.append [ $"(*, {mergeHistogram |> Seq.sumBy (fun pair -> pair.Value)})" ]
       |> String.join " "
 
-    logDebug $"Total buckets: {bucketsLength}"
-    logDebug $"Suppressed buckets: {bucketsLowCount}"
-    logDebug $"Suppressed buckets below threshold: {victimBucketsLength}"
+    logDebug $"Total buckets: {totalBuckets}"
+    logDebug $"Suppressed buckets: {suppressedBuckets}"
     logDebug $"Buckets with isolating column(s): {bucketsIsolatingColumn}"
+    logDebug $"Buckets with unknown column(s): {bucketsUnknownColumn}"
     logDebug $"Merged buckets: {bucketsMerged}"
     logDebug $"Merge distribution: {mergeHistogramStr}"
-    logDebug $"Total comparisons: {comparisons}"
-    logDebug $"Runtime: {stopwatch.Elapsed}"
 
-let private led (aggregationContext: AggregationContext) (buckets: Bucket array) (isolatingColumns: int array) =
-  let diagnostics = LedDiagnostics()
+type private CacheKey = Row
 
+/// Determines victim buckets and builds lookup caches.
+let private prepareBuckets (aggregationContext: AggregationContext) (buckets: Bucket array) =
+  let stopwatch = startStopwatch ()
   let lowCountIndex = Utils.lowCountIndex aggregationContext
   let groupingLabelsLength = aggregationContext.GroupingLabels.Length
+  let cacheKeySize = groupingLabelsLength - 1
+
+  let slice skipAt source =
+    let target = Array.zeroCreate cacheKeySize
+    Array.blit source 0 target 0 skipAt
+    Array.blit source (skipAt + 1) target skipAt (cacheKeySize - skipAt)
+    target
+
+  // For each column, we build a cache where we group buckets by their labels EXCLUDING that column.
+  // This means that every cache will associate siblings where the respective column is different.
+  let caches =
+    Array.init groupingLabelsLength (fun _i -> Dictionary<Row, MutableList<Bucket * bool>>(Row.equalityComparer))
+
+  let victimBuckets =
+    buckets
+    |> Array.choose (fun bucket ->
+      let lowCount = Utils.isLowCount lowCountIndex bucket
```

Comment:
I needed to spend some time grokking this, so let me rephrase what it does, and you tell me if I'm on the track:

We have a single pass through buckets here, but we do two fairly separate things:
1/ choose lowCount buckets to become victimBuckets
2/ populate a "two tier" mapping <column index> -> <bucket values excluding that index> -> <that bucket>. So we effectively cache which buckets would be potential siblings for a given "n-1 element subrows".

If that's more or less the case, maybe there's a way to make it a "two stage but effectively still single pass" like

```
buckets
|> Seq.map(fun which populates caches)
|> Array.choose(fun which chooses lowCount bucket tuples)
```

If those two "tasks" are truly independent, like it seems to me, then this would read easier, and you could provide one-liner comments as to what does each step do, this would make it even easier.

You'd need to carry over the `cacheKeys` from 1st stage to the 2nd, but maybe it's not a big deal?

Reply:
This loop is the most expensive part of the hook. I experimented with parallelization but gains are minimal since there will be contention for dictionary writes (which are not thread safe). Switching to `ConcurrentDictionary` didn't help either.

I agree that the function body is a little dirty with multiple side effects, but I'd favor performance over looks since this is a hot spot.

Reply:
but would the split as suggested hinder performance? I see that the reason for a single pass is performance, but I think the split would keep the same complexity. Hm maybe the loss would lie with the extra `cacheKeys` passed around.

So nevermind then. I'm out of ideas how to make it easier to grasp :(.

Reply:
Hah, I am about to make it more complex! I have some ideas with parallelization. If they don't work, I will try to simplify.

Note that there is already a bit of simplification because we no longer return cache keys but just the reference to the (mutable) list of siblings.